ciao-controller/client.go

```diff
@@ -73,6 +73,21 @@ func (client *ssntpClient) EventNotify(event ssntp.Event, frame *ssntp.Frame) {
 			glog.Warning("Error unmarshalling InstanceDeleted")
 			return
 		}
+
+		attachments, _ := client.ctl.ds.GetStorageAttachments(event.InstanceDeleted.InstanceUUID)
+		for _, attachment := range attachments {
+			if attachment.Ephemeral {
+				err = client.ctl.ds.DeleteBlockDevice(attachment.BlockID)
+				if err != nil {
+					glog.Warningf("Error deleting block device from datastore: %v", err)
+				}
+				err = client.ctl.DeleteBlockDevice(attachment.BlockID)
+				if err != nil {
+					glog.Warningf("Error deleting block device: %v", err)
+				}
+			}
+		}
+
 		client.ctl.ds.DeleteInstance(event.InstanceDeleted.InstanceUUID)
```

Comment:
I would have put the code to delete the block device into the where we update the attachments in datastore.go. Otherwise, you'll get to this section:
// update the state of the volume.
                        bd.State = types.Available
                        err = ds.UpdateBlockDevice(bd)
                        if err != nil {
                                glog.Warning(err)
                        }

And if the block device was deleted, you'll have issues.